pkg/controller/direct/directbase/directbase_controller.go

```diff
@@ -53,22 +55,81 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/source"
 )
 
-// Add creates a new controller for reconciling objects of the specified GVK, delegating actual resource reconciliation to the provided Model.
-func Add(mgr manager.Manager, gvk schema.GroupVersionKind, model Model, opts Deps) error {
+var ControllerBuilder directControllerBuilder
+
+func init() {
+	ControllerBuilder = directControllerBuilder{}
+}
+
+type directControllerBuilder struct {
+	modelMapper map[schema.GroupVersionKind]func(*controller.Config) Model
+}
+
+func (c *directControllerBuilder) RegisterModel(gvk schema.GroupVersionKind, modelFn func(*controller.Config) Model) {
+	if c.modelMapper == nil {
+		c.modelMapper = map[schema.GroupVersionKind]func(*controller.Config) Model{}
+	}
+	c.modelMapper[gvk] = modelFn
+}
+
+func (c *directControllerBuilder) AddController(mgr manager.Manager, config *controller.Config, gvk schema.GroupVersionKind, deps Deps) error {
 	immediateReconcileRequests := make(chan event.GenericEvent, k8s.ImmediateReconcileRequestsBufferSize)
 	resourceWatcherRoutines := semaphore.NewWeighted(k8s.MaxNumResourceWatcherRoutines)
-	reconciler, err := NewReconciler(mgr, immediateReconcileRequests, resourceWatcherRoutines, gvk, model, opts.JitterGenerator)
+
+	reconciler, err := c.NewReconciler(mgr, config, immediateReconcileRequests, resourceWatcherRoutines, gvk, deps.JitterGenerator)
 	if err != nil {
 		return err
 	}
 	return add(mgr, reconciler)
 }
 
-// NewReconciler returns a new reconcile.Reconciler.
-func NewReconciler(mgr manager.Manager, immediateReconcileRequests chan event.GenericEvent, resourceWatcherRoutines *semaphore.Weighted,
-	gvk schema.GroupVersionKind, model Model, jg jitter.Generator) (*DirectReconciler, error) {
+func (c *directControllerBuilder) IsDirectByCRD(crd *apiextensions.CustomResourceDefinition) bool {
+	for gvk, _ := range c.modelMapper {
+		if gvk.Group == crd.Spec.Group && gvk.Kind == crd.Spec.Names.Kind {
+			for _, version := range crd.Spec.Versions {
+				if gvk.Version == version.Name {
+					return true
+				}
+			}
+
+		}
+	}
+	return false
+}
+
+func (c *directControllerBuilder) IsDirectByGK(gk schema.GroupKind) bool {
+	if c.modelMapper == nil {
```

Comment:
FYI, you can `range` over a nil map; it is treated like an empty map.  (You can also read from it, you just can't write to it)

Reply:
oh, thanks! I always check the map nil first and never realized it can be iterated. 🤣